### PR TITLE
Add reason to WebSocket closure

### LINKED ISSUE #991: [Feature] Reason for websocket.close
I've recently started actively developing with websockets in starlette. When closing a websocket you can provide a code e.g. 1000, 1008 … to tell the client the basic reason. But the websocket protocol allows the developer to also give UTF-8-encoded data as a complementary description/ reason. This would be useful in multiple cases such as mentioning the specific violitioan for code 1008 (POLICY VIOLIATION). [Websockets](https://websockets.readthedocs.io/en/stable/api.html?highlight=reason#websockets.protocol.WebSocketCommonProtocol.close) supports this too and their [implementation](https://websockets.readthedocs.io/en/stable/_modules/websockets/protocol.html#WebSocketCommonProtocol.close) could maybe help.

Specification: [https://tools.ietf.org/](https://tools.ietf.org/html/rfc6455#section-5.5.1)

### PR DESCRIPTION
**Pull request by aminalaee**
Closes #991.

According to the ASGI spec [here](https://asgi.readthedocs.io/en/latest/specs/www.html#close-send-event) `reason` can be included in `websocket.close` which is an optional string.

<del>I think what @Kludex mentioned [here](https://github.com/encode/starlette/pull/1339#pullrequestreview-812914693) makes sense:</del>

<del>- We can include `code` in `websocket.close` only when it's specified in Starlette, and don't include it when not set, as per the ASGI spec it is optional int and already handled by Uvicorn [here](https://github.com/encode/uvicorn/blob/master/uvicorn/protocols/websockets/websockets_impl.py#L261) and Daphne [here](https://github.com/django/daphne/blob/main/daphne/ws_protocol.py#L262) and Hypercorn [here](https://gitlab.com/pgjones/hypercorn/-/blob/main/src/hypercorn/protocol/ws_stream.py#L293). </del>
- We will include `reason` in `websocket.close` only when it's specified as per the ASGI spec it's optional and will be handled by ASGI servers.

Any feedback would be appreciated.

### CODE REVIEW

File: tests/test_websockets.py
Comment (aminalaee) [inline tests/test_websockets.py]: This test is different from the above test, as that calls `close` on websocket but this one will await `WebSocketClose` manually. (For coverage)

File: docs/websockets.md
Comment (Kludex) [inline docs/websockets.md]:
```suggestion
* `await websocket.close(code=1000, reason=None)`
```

File: starlette/websockets.py
Comment (Kludex) [inline starlette/websockets.py]:
```suggestion
    def __init__(self, code: int = 1000, reason: str = None) -> None:
```

File: starlette/websockets.py
Comment (Kludex) [inline starlette/websockets.py]:
```suggestion
        message = {"type": "websocket.close", "code": code}

```

File: starlette/websockets.py
Comment (Kludex) [inline starlette/websockets.py]:
```suggestion
    def __init__(self, code: int = 1000, reason: str = None) -> None:
```

File: starlette/websockets.py
Comment (Kludex) [inline starlette/websockets.py]:
```suggestion
        message = {"type": "websocket.close", "code": code}

```

**Review by Kludex — COMMENTED** [review]
I understand the default value change on `code`, but that is another issue, and it's also considered a breaking change.

File: starlette/websockets.py
Comment (Kludex) [inline starlette/websockets.py]: Should we do this?

```suggestion
    await WebSocketClose(code, reason)
```

File: tests/test_websockets.py
Comment (Kludex) [reply tests/test_websockets.py]: This doesn't look cool. For this case, I guess we'd want to test the `WebSocketClose` class standalone, but see my comment above.

File: starlette/websockets.py
Comment (aminalaee) [reply starlette/websockets.py]: That makes sense but it should be something like:

```python
websocket_close = WebSocketClose(code, reason)
await websocket_close(self.scope, self.receive, self.send)
```

Comment (aminalaee) [reply starlette/websockets.py]: What do you think about this?

Comment (Kludex) [reply starlette/websockets.py]: Hmm, I think it's better to ignore my comment. :+1: 

The way it is already, it's easier to understand.

File: tests/test_websockets.py
Comment (Kludex) [reply tests/test_websockets.py]: But this should probably be tested elsewhere?

Comment (aminalaee) [reply tests/test_websockets.py]: I don't think we need the standalone anymore, thanks to your suggestion. I say let's drop it when there's no value in it.

Comment (aminalaee) [reply tests/test_websockets.py]: The only place `WebSocketClose` is used is in the `routing` where we close the connection. And in those cases we rely on the defaults. So the case with `reason` is not covered.

File: starlette/websockets.py
Comment (aminalaee) [reply starlette/websockets.py]: Actually when I first did this I thought that maybe we should get rid of `WebSocketClose`. I know it's out of scope and backwards incompatible.

But what you suggest is even better, it will make use of `WebSocketClose`, as I think it was meant to.
If you don't have any strong feelings about it I say let's wait for someone else's feedback here.

File: starlette/websockets.py
Comment (Kludex) [inline starlette/websockets.py]: image

Comment (Kludex) [reply starlette/websockets.py]: `reason = reason or ""`

File: starlette/testclient.py
Comment (Kludex) [inline starlette/testclient.py]: I think this should default to empty string, but at the end it doesn't make a difference. Your call.
```suggestion
            raise WebSocketDisconnect(message.get("code", 1000), message.get("reason", ""))
```

**Review by Kludex — APPROVED** [review]
Added a comment, but your call.

Everything is cool :+1: Thanks! :tada:

File: starlette/testclient.py
Comment (aminalaee) [reply starlette/testclient.py]: I noticed that earlier and changed `WebsocketDisconnect`:

```python
class WebSocketDisconnect(Exception):
     def __init__(self, code: int = 1000, reason: str = None) -> None:
         self.code = code
         self.reason = reason or ""
```

I think this will do that, right?

Comment (Kludex) [reply starlette/testclient.py]: Yeah, but the raise on the line that I've commented sends either the reason or None. In theory, `WebSocketDisconnect` doesn't receive `None` (looking at the annotation).

Comment (Kludex) [reply starlette/testclient.py]: The annotation is right.

Comment (aminalaee) [reply starlette/testclient.py]: Sorry I want to ask this out of curiosity, not for this PR.I get your point that in this internal code we will fix this.
But I guess both Python and MyPy (strict) will be ok about passing `None` to `reason`, right?

Comment (Kludex) [reply starlette/testclient.py]: If you mean the `strict` flag, then no. See below:

```bash
❯ mypy --strict main.py      
main.py:1: error: Incompatible default for argument "b" (default has type "None", argument has type "str")
main.py:4: error: Argument 1 to "a" has incompatible type "None"; expected "str"
Found 2 errors in 1 file (checked 1 source file)
(base) 
~/Development via 🐍 v3.8.8 on ☁️  marcelotryle@gmail.com 
❯ cat main.py                                                          
def a(b: str = None) -> None:
    return None

a(None)
```

Comment (Kludex) [reply starlette/testclient.py]: (we don't use the strict flag on starlette, jfyk)

Comment (aminalaee) [reply starlette/testclient.py]: But this will be ok:

```python
import typing

def a(b: typing.Optional[str] = None) -> None:
    return None

a(None)
```

I'll fix this as it's not probably related and we can discuss later. Thank you.

**Pull request merged**
